pkg/kv/kvserver/gc/gc_random_test.go

```diff
@@ -183,6 +192,273 @@ func BenchmarkRun(b *testing.B) {
+			},
+			// GC everything beyond intent resolution threshold
+			ttlSec: int32(intentAgeThreshold.Seconds()),
+		},
+		{
+			ds: someVersionsMidSizeRows,
+			now: hlc.Timestamp{
+				WallTime: 100 * time.Second.Nanoseconds(),
+			},
+			ttlSec: 1,
+		},
+	} {
+		t.Run(fmt.Sprintf("%v@%v,ttl=%vsec", tc.ds, tc.now, tc.ttlSec), func(t *testing.T) {
+			rng := rand.New(rand.NewSource(1))
+			eng := storage.NewDefaultInMemForTesting()
+			defer eng.Close()
+
+			tc.ds.dist(N, rng).setupTest(t, eng, *tc.ds.desc())
+			snap := eng.NewSnapshot()
+
+			// Run GCer over snapshot.
+			ttl := time.Duration(tc.ttlSec) * time.Second
+			gcThreshold := CalculateThreshold(tc.now, ttl)
+			intentThreshold := tc.now.Add(-intentAgeThreshold.Nanoseconds(), 0)
+
+			gcer := makeFakeGCer()
+			gcInfoNew, err := Run(ctx, tc.ds.desc(), snap, tc.now,
+				gcThreshold, RunOptions{IntentAgeThreshold: intentAgeThreshold}, ttl,
+				&gcer,
+				gcer.resolveIntents,
+				gcer.resolveIntentsAsync)
+			require.NoError(t, err)
+
+			// Handle GC + resolve intents.
+			var stats enginepb.MVCCStats
+			require.NoError(t,
+				storage.MVCCGarbageCollect(ctx, eng, &stats, gcer.requests(), gcThreshold))
+			for _, i := range gcer.intents {
+				l := roachpb.LockUpdate{
+					Span:   roachpb.Span{Key: i.Key},
+					Txn:    i.Txn,
+					Status: roachpb.ABORTED,
+				}
+				_, err := storage.MVCCResolveWriteIntent(ctx, eng, &stats, l)
+				require.NoError(t, err, "failed to resolve intent")
+			}
+
+			assertGCInvariants(t, eng, snap, *tc.ds.desc(), tc.now, gcThreshold, intentThreshold, ttl,
+				gcInfoNew)
+		})
+	}
+}
+
+// assertGCInvariants will create a stream of expected values based on full data
+// set contained in provided "before" reader and compare it with the "after"
+// reader that contains data after applying GC request.
+func assertGCInvariants(
+	t *testing.T,
+	after, before storage.Reader,
+	desc roachpb.RangeDescriptor,
+	now, gcThreshold, intentThreshold hlc.Timestamp,
+	gcTTL time.Duration,
+	gcInfo Info,
+) {
+	failureDetails := func(key storage.MVCCKey) string {
+		return fmt.Sprintf("key=%s, GC time=%s, intent time=%s, key history=%s", key.String(),
+			gcThreshold.String(), intentThreshold.String(), getKeyHistory(t, before, key.Key))
+	}
+
+	// Validation works on original data applying simple GC eligibility to key
+	// history and only returning remaining elements and accounting for remaining
+	// ones as gc.Info.
+	expInfo := Info{
+		Now:       now,
+		GCTTL:     gcTTL,
+		Threshold: gcThreshold,
+	}
+	pointIt := before.NewMVCCIterator(storage.MVCCKeyAndIntentsIterKind,
+		storage.IterOptions{
+			LowerBound: desc.StartKey.AsRawKey(),
+			UpperBound: desc.EndKey.AsRawKey(),
+			KeyTypes:   storage.IterKeyTypePointsOnly,
+		})
+	defer pointIt.Close()
+	pointIt.SeekGE(storage.MVCCKey{Key: desc.StartKey.AsRawKey()})
+	pointExpectations := getPointExpectations(t, pointIt, gcThreshold, intentThreshold, &expInfo)
```

Comment:
It's initially not clear that this returns a generator function that closes over the passed parameters -- consider renaming the functions to make this clearer.